multicluster/cmd/service-mirror/main.go

```diff
@@ -84,79 +103,159 @@ func Main(args []string) {
+					cleanupWorkers()
+					return
+				case event, ok := <-results:
+					if !ok {
+						log.Info("Link watch terminated; restarting watch")
+						loop = false
+						break
+					}
+					switch obj := event.Object.(type) {
+					case *dynamic.Unstructured:
+						if obj.GetName() == linkName {
+							switch event.Type {
+							case watch.Added, watch.Modified:
+								link, err := multicluster.NewLink(*obj)
+								if err != nil {
+									log.Errorf("Failed to parse link %s: %s", linkName, err)
+									continue
+								}
+								log.Infof("Got updated link %s: %+v", linkName, link)
+								creds, err := loadCredentials(ctx, link, *namespace, k8sAPI)
+								if err != nil {
+									log.Errorf("Failed to load remote cluster credentials: %s", err)
+								}
+								err = restartClusterWatcher(ctx, link, *namespace, creds, controllerK8sAPI, *requeueLimit, *repairPeriod, metrics, *enableHeadlessSvc)
+								if err != nil {
+									// failed to restart cluster watcher; give a bit of slack
+									// and restart the link watch to give it another try
+									log.Error(err)
+									time.Sleep(linkWatchRestartAfter)
+									linkWatch.Stop()
+								}
+							case watch.Deleted:
+								log.Infof("Link %s deleted", linkName)
+								cleanupWorkers()
+							default:
+								log.Infof("Ignoring event type %s", event.Type)
 							}
-						default:
-							log.Infof("Ignoring event type %s", event.Type)
 						}
+					default:
+						log.Errorf("Unknown object type detected: %+v", obj)
 					}
-				default:
-					log.Errorf("Unknown object type detected: %+v", obj)
 				}
 			}
 		}
 	}
+
+	hostname, found := os.LookupEnv("HOSTNAME")
+	if !found {
+		log.Fatal("Failed to fetch 'HOSTNAME' environment variable")
+	}
+
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("service-mirror-write-%s", linkName),
+			Namespace: *namespace,
+			Labels: map[string]string{
+				"component":                      "linkerd-service-mirror",
+				"mirror.linkerd.io/cluster-name": linkName,
+			},
+		},
+		Client: k8sAPI.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: hostname,
+		},
+	}
+
+	for loop := true; loop; {
+		// RunOrDie will block until a lease is acquired. Once a lease is
+		// acquired, the main watcher loop will start a Link watch. If the lease
+		// is lost, clean-up watchers and re-attempt to acquire.
+		leaderelection.RunOrDie(rootCtx, leaderelection.LeaderElectionConfig{
+			// When runtime context is cancelled, lock will be released. Implies any
+			// code guarded by the least _must_ finish before cancelling.
+			ReleaseOnCancel: true,
+			Lock:            lock,
+			LeaseDuration:   LEASE_DURATION,
+			RenewDeadline:   LEASE_RENEW_DEADLINE,
+			RetryPeriod:     LEASE_RETRY_PERIOD,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(ctx context.Context) {
+					// Rely on cancellation to be propagated from leader elector
```

Comment:
This is the crux of how we stop doing work when we lose the lease, and it's a bit subtle, so I think it's worth calling more attention to this and being a bit more explicit in the comment.  When we lose the lease, `RunOrDie` will cancel this context which, in turn, will cause us to cancel all the work in the `run` function.